src/shogun/base/AnyParameter.h

```diff
@@ -21,49 +27,102 @@ namespace shogun
+	 * vectors and bias
+	 */
 	class AnyParameterProperties
 	{
 	public:
-		AnyParameterProperties()
-		    : m_description(), m_model_selection(MS_NOT_AVAILABLE),
-		      m_gradient(GRADIENT_NOT_AVAILABLE)
+		static const int32_t HYPERPARAMETER = 0x00000001;
+		static const int32_t GRADIENT_PARAM = 0x00000010;
+		static const int32_t MODEL_PARAM = 0x00000100;
+
+		/** default constructor where all parameter types are switched to false
+		 */
+		AnyParameterProperties() : m_description(), m_mask_attribute(0x00000000)
 		{
 		}
+		/** legacy constructor */
 		AnyParameterProperties(
 		    std::string description,
 		    EModelSelectionAvailability model_selection = MS_NOT_AVAILABLE,
-		    EGradientAvailability gradient = GRADIENT_NOT_AVAILABLE)
+		    EGradientAvailability gradient = GRADIENT_NOT_AVAILABLE,
+		    bool hyperparameter = false)
 		    : m_description(description), m_model_selection(model_selection),
 		      m_gradient(gradient)
 		{
+			m_mask_attribute = 0x00000000;
+			if (model_selection)
+				m_mask_attribute |= MODEL_PARAM;
+			if (gradient)
+				m_mask_attribute |= GRADIENT_PARAM;
+			if (hyperparameter)
+				m_mask_attribute |= HYPERPARAMETER;
+		}
+		/** mask constructor */
+		AnyParameterProperties(std::string description, int32_t mask_attribute)
+		    : m_description(description)
+		{
+			m_mask_attribute = mask_attribute;
 		}
+		/** copy contructor */
 		AnyParameterProperties(const AnyParameterProperties& other)
 		    : m_description(other.m_description),
 		      m_model_selection(other.m_model_selection),
-		      m_gradient(other.m_gradient)
+		      m_gradient(other.m_gradient),
+		      m_mask_attribute(other.m_mask_attribute)
 		{
 		}
-
+		/** description getter */
 		std::string get_description() const
 		{
 			return m_description;
 		}
-
+		/** model selection flag getter */
 		EModelSelectionAvailability get_model_selection() const
 		{
-			return m_model_selection;
+			EModelSelectionAvailability return_val;
+			if (m_mask_attribute & HYPERPARAMETER)
+				return_val = EModelSelectionAvailability::MS_AVAILABLE;
+			else
+				return_val = EModelSelectionAvailability::MS_NOT_AVAILABLE;
+			return return_val;
 		}
 
+		/** gradient flag getter */
 		EGradientAvailability get_gradient() const
 		{
-			return m_gradient;
+			EGradientAvailability return_val;
+			if (m_mask_attribute & GRADIENT_PARAM)
+				return_val = EGradientAvailability::GRADIENT_AVAILABLE;
+			else
+				return_val = EGradientAvailability::GRADIENT_NOT_AVAILABLE;
+			return return_val;
+		}
+
+		/** hyperparameter flag getter */
+		bool get_hyperparameter() const
+		{
+			bool return_val;
+			if (m_mask_attribute & HYPERPARAMETER)
+				return_val = true;
+			else
+				return_val = false;
+			return return_val;
 		}
 
 	private:
 		std::string m_description;
 		EModelSelectionAvailability m_model_selection;
 		EGradientAvailability m_gradient;
+		/** mask to store all param flags*/
+		int32_t m_mask_attribute;
```

Comment:
what about calling it "attribute_mask" ? :)

Reply:
Yup, I just noticed that I switched the order...